models/parrot_rate_neuron.h

```diff
@@ -0,0 +1,287 @@
+/*
+ *  parrot_rate_neuron.h
```

Comment:
I find the name 'parrot_rate_neuron' a bit misleading.
In NEST, a parrot neuron is one that exactly repeats what it gets in the input.
Here you apply a nonlinear transformation. So I would rather think of it as a rate neuron with time constant tau=0.
tau=0 is currently not possible in the implementation of rate_neuron_ipn etc.
But would it not be possible to insert some if statements in the definitions of the propagators of rate_neuron_ipn to enable tau=0? Then you could spare this new model here. Although it might be also good to have these two cases, tau>0  and tau=0, separate as in the former case the neuron has some internal dynamics and in the latter case not.
If you decide for keeping this new model, I would suggest to rename it to something like 'rate_nonlinearity'.

Also keep in mind that the behavior for linear_summation=True is different here from rate_neuron_ipn/opn because here you do not keep excitatory and inhibitory inputs separate.

Reply:
You are right, that the `parrot_rate_neuron` only repeats its input in the linear version, i.e. `lin_rate_parrot`. All other versions do indeed apply the corresponding nonlinearity to the input.
I like the name as it gives a hint what the neuron is used for, which is basicly the buffering (and here also the modifcation) of its received input. Anyway if @heplesser also agrees that the name is misleading or misplaced here we could indeed change it.

You are also right that the class basicly represents and behaves like a `rate_neuron_ipn` with `tau=0`. As far as I remember the main reasons why we implemented this in a seperate class (besides the reason that the integration into `rate_neuron_ipn` would require a couple of new if-statements, which would make the code even more complex) is that it makes simulation scripts much easier to read: It is much more intuitive to think of those neurons as buffer or helper neurons than to classify them as `rate_neuron_ipn` neurons with `tau=0`. Did I get this right @MSenden?

Reply:
@janhahne yes, you are right. In the model whose implementation drove my choice for rate neuron types, some neurons would apply different nonlinearities to inputs originating from different sources. To handle this the outputs of the different sources are passed through parrot neurons applying the desired nonlinearities before they reach a linear rate neuron. @ddahmen You are also right that this is equivalent as using a rate neuron with the desired nonlinearity and setting tau to zero but when building a model I was under the impression that the code is easier to follow if a source's output is passed through a helper unit explicitly dedicated to instantaneously applying a nonlinearity than having to recognize another neuron as doing precisely this based on its parameter settings.

However, there is indeed no use-case for `lin_rate_parrot` as it does not apply a nonlinearity

Reply:
I agree that a model is much easier to understand if the 'parrot_rate_neuron' is not merged with the other types of neurons. So only the naming problem remains. What do you think @heplesser ?

Reply:
@janhahne @MSenden @ddahmen Thank you for discussing this point in detail! If I understand the discussion correctly, then you use the "parrots" to transform rate signals on their way from a sender neuron to a target neuron. You implement this transformation as a separate network node to avoid complexities in input handling in the target neuron. This seems sensible, but I agree with @ddahmen that the naming is unfortunate, as a parrot should indeed repeat its input perfectly. I am also a bit confused by the use of the term "buffer" in the discussion, as my understanding is that we are here talking about instantaneous transformations.

I would suggest `rate_transformer_<func>` as name, where `<func>` describes the transformation function applied. If the linear version does nothing, I would drop it, but otherwise I would agree to provide all possible transformeres if adding them only requires an additional template instantiation.

Reply:
I like your suggestion and implemented the change of the model name(s). Using the term "buffer" in the discussion was indeed unfortunate - i meant a "helper" neuron.